File: src/main/java/raylras/zen/model/type/NumberType.java
package raylras.zen.model.type;

import raylras.zen.model.CompilationEnvironment;
import raylras.zen.model.SymbolProvider;
import raylras.zen.model.symbol.Operator;
import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolFactory;

import java.util.List;
import java.util.function.UnaryOperator;

public abstract class NumberType extends Type implements SymbolProvider {

    @Override
    public boolean isInheritedFrom(Type type) {
        if (type instanceof NumberType) {
            return true;
        }
        return super.isInheritedFrom(type);
    }

    @Override
    public boolean isCastableTo(Type type, CompilationEnvironment env) {
        if (type == StringType.INSTANCE) {
            return true;
        }
        return super.isCastableTo(type, env);
    }

    @Override
    public List<Symbol> getSymbols() {
        return SymbolFactory.builtinSymbols()
                .operator(Operator.ADD, this, params -> params.parameter("val", this))
                .operator(Operator.SUB, this, params -> params.parameter("val", this))
                .operator(Operator.MUL, this, params -> params.parameter("val", this))
                .operator(Operator.DIV, this, params -> params.parameter("val", this))
                .operator(Operator.MOD, this, params -> params.parameter("val", this))
                .operator(Operator.CAT, StringType.INSTANCE, params -> params.parameter("str", StringType.INSTANCE))
                .operator(Operator.COMPARE, IntType.INSTANCE, params -> params.parameter("val", this))
                .operator(Operator.NEG, this, UnaryOperator.identity())
                .build();
    }
}

File: core/src/main/kotlin/raylras/intellizen/type/NumberType.kt
package raylras.intellizen.type

import raylras.intellizen.CompilationEnvironment
import raylras.intellizen.symbol.Operator
import raylras.intellizen.symbol.Symbol
import raylras.intellizen.symbol.SymbolProvider
import raylras.intellizen.symbol.symbolSequence

interface NumberType : Type, SymbolProvider {
    override fun isCastableTo(that: Type?, env: CompilationEnvironment): Boolean {
        return when (that) {
            is NumberType, is StringType -> {
                true
            }

            else -> super.isCastableTo(that, env)
        }
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        val builtin = symbolSequence {
            operator(Operator.NEG, this@NumberType)
            operator(Operator.ADD, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.SUB, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.MUL, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.DIV, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.MOD, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.CONCAT, StringType) {
                parameter("value", StringType)
            }
            operator(Operator.EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.NOT_EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.LESS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.LESS_EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.GREATER, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.GREATER_EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
        }
        val classDeclared = env?.classes
            ?.firstOrNull { it.qualifiedName == IntType.typeName }
            ?.getSymbols()
            ?: emptySequence()
        return builtin + classDeclared + getExpands(env)
    }
}
